src/Controller/Controller.php

```diff
@@ -757,12 +758,79 @@ public function render(?string $template = null, ?string $layout = null): Respon
         if ($builder->getTemplate() === null) {
             $builder->setTemplate($this->request->getParam('action'));
         }
+        $viewClass = $this->chooseViewClass();
+        $view = $this->createView($viewClass);
 
-        $view = $this->createView();
         $contents = $view->render();
-        $this->setResponse($view->getResponse()->withStringBody($contents));
+        $response = $view->getResponse()->withStringBody($contents);
 
-        return $this->response;
+        $viewContentType = $view->getContentType();
+        $responseType = $response->getHeaderLine('Content-Type');
+        if ($viewContentType && ($responseType === '' || substr($responseType, 0, 9) === 'text/html')) {
+            $response = $response->withHeader('Content-Type', $viewContentType);
+        }
+
+        return $this->setResponse($response)->response;
+    }
+
+    /**
+     * Get the list of View classes this controller can negotiate with.
+     *
+     * Each view class must implement the `getContentType()` hook method
+     * to participate in negotiation.
+     *
+     * @see Cake\Http\ContentTypeNegotiation
+     * @return array<string>
+     */
+    public function getViewClasses(): array
+    {
+        return [];
+    }
+
+    /**
+     * Use the view classes defined on this controller to view
+     * selection based on content-type negotiation.
+     *
+     * TODO: Should this also consider $request->getParam('_ext') as well?
+     *
+     * @return string|null The chosen view class or null for no decision.
+     */
+    protected function chooseViewClass(): ?string
+    {
+        $possibleViewClasses = $this->getViewClasses();
+        if (empty($possibleViewClasses)) {
+            return null;
+        }
+        if ($this->viewBuilder()->getClassName() !== null) {
+            return null;
+        }
+
+        $typeMap = [];
+        foreach ($possibleViewClasses as $class) {
+            $viewContentType = $class::getContentType();
+            if ($viewContentType && !isset($typeMap[$viewContentType])) {
+                $typeMap[$viewContentType] = $class;
+            }
+        }
+        $request = $this->getRequest();
+
+        // Prefer the _ext route parameter if it is defined.
+        $ext = $request->getParam('_ext');
+        if ($ext) {
+            $extType = $this->response->getMimeType($ext);
+            if (isset($typeMap[$extType])) {
+                return $typeMap[$extType];
+            }
+        }
+
+        // Use accept header based negotiation.
+        $contentType = new ContentTypeNegotiation();
+        $preferredType = $contentType->preferredType($request, array_keys($typeMap));
+        if (!$preferredType) {
+            return null;
```

Comment:
I am wondering if we should throw a `NotAcceptableException` exception here. For e.g. if my API only supports JSON and some uses `Accept` to ask for another type then returning a `406 Not Acceptable` response seems reasonable.

Reply:
We would have already run the action though. I think an exception could work well if we triggered it before the controller's action was called.

Reply:
True, if there's an exception it should be before the action runs.